src/Database/Esqueleto/Internal/Internal.hs

```diff
@@ -1647,7 +1651,13 @@ unsafeSqlBinOp op (ERaw p1 f1) (ERaw p2 f2) = ERaw Parens f
                  (b2, vals2) = f2 info
              in ( parensM p1 b1 <> op <> parensM p2 b2
                 , vals1 <> vals2 )
-unsafeSqlBinOp _ _ _ = throw (CompositeKeyErr SqlBinOpError)
+unsafeSqlBinOp op a b = unsafeSqlBinOp op (construct a) (construct b)
+    where construct :: SqlExpr (Value a) -> SqlExpr (Value a)
+          construct (ERaw p f)        = ERaw Never $ \info ->
+            let (b1, vals) = f info
+             in (parensM p b1, vals)
+          construct (ECompositeKey f) =
+            ERaw Parens $ \info -> (uncommas $ f info, mempty)
```

Comment:
:tada: 

This looks good to me! Can we write a test for it?

I think something like:

```haskell
select $
from $ \p -> do
where_ (p ^. SomeCompositeField `between` (val $ MkCompositeField 1 2, val $ MkCompositeField 3 4)
```

would do it.

Reply:
I added a test for it [here](https://github.com/bitemyapp/esqueleto/blob/34d3805ce0d0c15b834975cebde898eeff64a65e/test/Common/Test.hs#L655). Is that enough?

Reply:
You should be able to do `val $ PointKey 1 2` instead of manually calling the `ECompositeKey` constructor

Reply:
That would lead to failing a failing test as we discussed before, due to the fact that `val` wraps everything on `ERaw` ADT :disappointed:

Reply:
I thought your change to `unsafeSqlBinOp` fixed that?

Reply:
Not really, with this change `unsafeSqlBinOp` can know handle composite keys but you still have to manually specify it as composite as on the test, with `val` at its current state is impossible to create `ECompositeKey` from any `persistent` composite primary key.

Reply:
:thinking: 

Users should be able to do this seamlessly - comparing a projected composite key and a `val` composite key should work. So I'm curious what happens now when you run your code with the `val` in a `between` function

Reply:
When using scalar values only it works, for only composite values works two, the problem here is when you use a mixture of both and when using `val` it transforms everything on scalar values or `ERaw`, that's more of an issue of `val` instead of `unsafeSqlBinOp` in my opinion, and as you mentioned [here](https://github.com/bitemyapp/esqueleto/pull/127#discussion_r298259357) at least is my understanding that users should be aware of this, that's why `unsafeSqlBinOp` will not `throw` any error but will fail at runtime with SQL syntax error

Reply:
Hey @parsonsmatt, have you been able to review the last comment or maybe came up with another idea for these changes?

Reply:
Thank you for pinging me again! I need to think about this for a bit and I've put a note on my todo list to check on it.

Also pinging @bitemyapp and maybe @FintanH to provide input if they have time/energy to do so

Reply:
I've got this cloned locally and I'm going to try writing some tests for expected behavior with `val $ CompositeKey 'a' 'b'` etc to see if I can finagle it into a working place